File: backend/services/crypto.py
import base64
import hashlib
import json
import os
from typing import Dict

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def derive_key_from_patient_address(patient_address: str) -> bytes:
    normalized_address = patient_address.lower().encode("utf-8")
    return hashlib.sha256(normalized_address).digest()


def encrypt_raw_text(raw_text: str, patient_address: str) -> Dict[str, str]:
    key = derive_key_from_patient_address(patient_address)
    iv = os.urandom(12)

    aesgcm = AESGCM(key)
    encrypted = aesgcm.encrypt(iv, raw_text.encode("utf-8"), None)
    ciphertext = encrypted[:-16]
    tag = encrypted[-16:]

    return {
        "alg": "AES-256-GCM",
        "iv": base64.b64encode(iv).decode("utf-8"),
        "ciphertext": base64.b64encode(ciphertext).decode("utf-8"),
        "tag": base64.b64encode(tag).decode("utf-8"),
    }


def decrypt_raw_text(encrypted_payload: Dict[str, str], patient_address: str) -> str:
    key = derive_key_from_patient_address(patient_address)
    iv = base64.b64decode(encrypted_payload["iv"])
    ciphertext = base64.b64decode(encrypted_payload["ciphertext"])
    tag = base64.b64decode(encrypted_payload["tag"])

    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext + tag, None)
    return plaintext.decode("utf-8")


def hash_encrypted_blob(encrypted_payload: Dict[str, str]) -> str:
    canonical_json = json.dumps(encrypted_payload, sort_keys=True, separators=(",", ":"))
    digest = hashlib.sha256(canonical_json.encode("utf-8")).hexdigest()
    return f"0x{digest}"

File: frontend/src/services/crypto.js
function base64ToUint8Array(base64) {
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);

  for (let i = 0; i < binary.length; i += 1) {
    bytes[i] = binary.charCodeAt(i);
  }

  return bytes;
}

function uint8ArrayToBase64(bytes) {
  let binary = "";
  bytes.forEach((value) => {
    binary += String.fromCharCode(value);
  });
  return btoa(binary);
}

async function deriveAesKeyFromAddress(patientAddress) {
  const normalized = patientAddress.toLowerCase();
  const digestBuffer = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder().encode(normalized)
  );

  return crypto.subtle.importKey("raw", digestBuffer, { name: "AES-GCM" }, false, [
    "encrypt",
    "decrypt",
  ]);
}

export async function encryptRawText(rawText, patientAddress) {
  const key = await deriveAesKeyFromAddress(patientAddress);
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const encryptedBuffer = await crypto.subtle.encrypt(
    {
      name: "AES-GCM",
      iv,
      tagLength: 128,
    },
    key,
    new TextEncoder().encode(rawText)
  );

  const encryptedBytes = new Uint8Array(encryptedBuffer);
  const ciphertext = encryptedBytes.slice(0, encryptedBytes.length - 16);
  const tag = encryptedBytes.slice(encryptedBytes.length - 16);

  return {
    alg: "AES-256-GCM",
    iv: uint8ArrayToBase64(iv),
    ciphertext: uint8ArrayToBase64(ciphertext),
    tag: uint8ArrayToBase64(tag),
  };
}

export async function decryptRawText(encryptedPayload, patientAddress) {
  const key = await deriveAesKeyFromAddress(patientAddress);
  const iv = base64ToUint8Array(encryptedPayload.iv);
  const ciphertext = base64ToUint8Array(encryptedPayload.ciphertext);
  const tag = base64ToUint8Array(encryptedPayload.tag);

  const combined = new Uint8Array(ciphertext.length + tag.length);
  combined.set(ciphertext, 0);
  combined.set(tag, ciphertext.length);

  const decryptedBuffer = await crypto.subtle.decrypt(
    {
      name: "AES-GCM",
      iv,
      tagLength: 128,
    },
    key,
    combined
  );

  return new TextDecoder().decode(decryptedBuffer);
}
